### PR TITLE
[lib/logical] Improve Excel parity for undefined, null & error inputs

### PR DESCRIPTION
**Pull request by pbadenski**
Scope: lib/logical.js

### CODE REVIEW

File: lib/logical.js [hunk 1]
```diff
@@ -4,9 +4,16 @@ var information = require('./information');
 
 exports.AND = function() {
   var args = utils.flatten(arguments);
+  if (args.every(function (arg) { return arg === undefined || arg === null; })) {
```
Comment (pbadenski) [inline lib/logical.js:7]: From @nicolashefti (orig: https://github.com/formulajs/formulajs/pull/51/files#r659165052):

you add two new loops on all arguments. Can you just move the tests inside the existing for loop?

File: lib/logical.js [hunk 2]
```diff
@@ -66,6 +82,13 @@ exports.NOT = function(logical) {
 
 exports.OR = function() {
   var args = utils.flatten(arguments);
+  if (args.every(function (arg) { return arg === undefined || arg === null; })) {
```
Comment (pbadenski) [inline lib/logical.js:85]: From @nicolashefti (orig: https://github.com/formulajs/formulajs/pull/51/files#r659165100)

Same as AND: You add two new loops on all arguments. Can you just move the tests inside the existing for loop?

File: lib/logical.js (hunk 1 above)
Comment (pbadenski) [reply lib/logical.js:7]: 1. I can move Lines 10-12 to inside the loop easily (function needs to short-circuit return an error if any argument is an error).
2. The best solution I came up for lines 7-9 looks like this - what do you think:
```javascript
  var result;
  for (var i = 0; i < args.length; i++) {
    if (args[i] === undefined || args[i] === null) {
      continue;
    }
    if (result === undefined) {
      result = true;
    }
    if (!args[i]) {
      result = false;
    }
  }
  if (result === undefined) {
    return error.value;
  }
  return result;
```

Comment (nicolashefti) [reply lib/logical.js:7]: @pbadenski  can't you just initialize result to error.value ? This way you do not need the last check.

Can you directly update the code with the overall solution (for both loops)?

File: test/logical.js [hunk 3]
```diff
@@ -5,6 +5,9 @@ var should = require('should');
 
 describe('Logical', function() {
   it('AND', function() {
+    logical.AND(undefined, undefined).should.equal(error.value);
+    logical.AND(undefined, true).should.equal(true);
+    logical.AND(error.na, true).should.equal(error.na);
```
Comment (nicolashefti) [inline test/logical.js:10]: In Excel AND returns true with number (unless false with 0), string are treated like an empty value. Can you add two other tests?

File: lib/logical.js (hunk 1 above)
Comment (pbadenski) [reply lib/logical.js:7]: that's a good idea, I'll do that and update pull request

File: test/logical.js (hunk 3 above)
Comment (pbadenski) [reply test/logical.js:10]: will do

**Review by nicolashefti — APPROVED** [review]
@pbadenski that looks good to me.

**Pull request merged**